rgw: Potential fix for possible 500 on POST

adamemerson opened the pull request:
Force re-fetching the bucket info if our cache gets out of date.

yehudasa commented on src/rgw/rgw_bucket.cc:
```diff
@@ -423,7 +423,22 @@ int rgw_bucket_set_attrs(RGWRados *store, RGWBucketInfo& bucket_info,
+  auto r = rgw_bucket_instance_store_info(store, key, bl, false,
+					  &attrs, objv_tracker, real_time());
+  if (r == -ECANCELED) {
+    for (auto i = 0; i < 15; ++i || r != -ECANCELED) {
```
@adamemerson could it be that the condition here is broken?

amitkumar50 replied:
correct it can be broken

yehudasa commented:
@adamemerson I understand why adding the `force` here fixes things for you, but it might be that we're just hiding the bug. Moreover, and looking at all the callers to `rgw_bucket_set_attrs()` it's not clear to me why this needs to be solved here and not at a higher level.
Unless it's a cache coherency bug (which after reading the code I have reason to believe it isn't), then the generic approach would be to retry the operation above, similar to what we do in `RGWBucketCreate::execute()`.
However, I have another question. Do we even want to force these attributes if we lost a race? There are a few different cases. If each of the racing operations overwrites completely the attributes, then the effect of retrying is pretty much the same as the effect of not doing anything and returning successfully. One of the operations succeeded, one failed, and since they were executing concurrently we can't and don't need to guarantee which is the one that needs to win.
When we're dealing with additive attributes we can't do that, but also doing a retry like in this PR isn't not the correct approach, as we'd end up overriding the racing attributes with older data. In this case we'd need to send the error back (as we did before), and at the caller level do a retry of the entire operation logic -- read the bucket attributes, merge the new attributes, try to write, and so on.
After inspection of the callers, it seems to me that in all the cases where we call `rgw_bucket_set_attrs()` we merge the existing attributes with new ones, so in all cases we need to retry the operation from the top level. It makes sense, as these attributes include things like bucket ACLs, CORS config, etc. that can be changed separately.

adamemerson commented:
@yehudasa I agree about higher level. I thought merging was happening lower than where I fixed it, but you are right.

As to the other stuff, the reason I think we /want/ to force, or at least invalidate, something is that in the hypothetical bug, someone says that RGW gets into a state where the same operation returns 500 until someone restarts it, so I expect cache is getting out of sync somewhere.

yehudasa commented:
@adamemerson I'd separate the two issues. Also, not sure I'm happy with forcing invalidate because we have a bug somewhere, I'd rather find and fix that bug.

adamemerson commented:
@yehudasa So would I. SO. Given that, would the case I described be something that could be the result of Watch/Notify traffic getting lost under heavy load? If there's supposed to be an update coming that the RGW in question never sees? That's my running theory, but I haven't been able to reproduce it. (Though if that's the case I'd suggest we should probably have some way of noticing when we're out to lunch and refreshing from the OSDs.)

yehudasa commented:
@adamemerson maybe something like watch/notify getting lost, but the watch/notify protocol supposed to identify these cases I think. That been said, it does make sense to invalidate cache if we get ECANCELED anyway. But we can then race with cache updates, so we should use the objv_tracker to make sure that we don't invalidate an updated cache entry.

adamemerson commented:
@yehudasa All right. So! I can hack something up tomorrow. Does the plan of:

Add a try_refresh_bucket_info function that rather than doing a FORCED update, compares the cached bucket version to the one supplied. If the cached one is higher, it just returns that. Otherwise it does a straight up update from the OSD.

And then have the operation's execute function use try_refresh_bucket on -ECANCELED and make sure it goes through merging the attributes and such properly?

adamemerson commented:
@yehudasa How's this? On -ECANCELLED the operation will run `try_refresh_bucket_info`

`try_refresh_bucket_info` will attempt to update from cache first and if the cache is still lagging behind, note that this is abnormal in the log and try to fetch it.

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11884,23 +11884,17 @@ int RGWRados::convert_old_bucket_info(RGWObjectCtx& obj_ctx,
 {
   bucket_info_entry e;
   string bucket_entry;
   rgw_make_bucket_entry_name(tenant, bucket_name, bucket_entry);
```
@adamemerson should pass bucket_entry as param, otherwise it's going to be calculated twice on cache miss.

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+    return 0;
+  }
+
+  lderr(cct) << "Beware! The bucket info cache is inconsistent. This is an "
```
change `Beware!` to `ERROR:`

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+  string bucket_entry;
+  rgw_make_bucket_entry_name(info.bucket.tenant, info.bucket.name, bucket_entry);
+
+  auto checkv = info.objv_tracker.read_version;
```
Note that there's a string copy here so it'd be better to take a reference. Maybe use reference here, and if this check fails only then assign it to a local variable?

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+             << "will try to recover." << dendl;
+
+  binfo_cache->invalidate(bucket_entry);
+  auto r = bucket_info_cache_miss(obj_ctx, info.bucket.tenant,
```
Not crazy about having the whole get_bucket_logic() twice. This one is pretty much a superset of the other one, maybe consider some kind of a logic merge (while keeping the same interfaces)?

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+                                  info.bucket.name, info, pmtime,
+                                  pattrs);
+
+  if (r != 0) {
```
should really check for `(r < 0)` as error

yehudasa commented on src/rgw/rgw_op.cc:
```diff
@@ -3982,55 +3982,67 @@ void RGWPutMetadataBucket::execute()
+					&s->bucket_attrs)) {
+      break;
+    }
+  } while (--tries_remaining);
```
so, the good news is that it seems like `RGWPutMetadataBucket::execute()` doesn't have any local variables. This means that instead of putting this whole block of code in a `do { } while`, you can put it in a separate method, and keep the `do { } while` here.

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+                                pattrs);
+}
+
+bool RGWRados::try_refresh_bucket_info(RGWBucketInfo& info,
```
make this return an error instead of a bool.

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+
+  if (r != 0) {
+    lderr(cct) << "Refresh failed with error: " << r << dendl;
+    return false;
```
`return r;`

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11967,6 +11961,73 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+                                  pattrs);
+
+  if (r != 0) {
+    lderr(cct) << "Refresh failed with error: " << r << dendl;
```
`ERROR: refreshed failed, error:` .. easier to scan in the logs

yehudasa commented:
@adamemerson see my comments. Logic is sound, just a few minor changes. Also, I think this to be applied anywhere else in rgw_op.cc where the issue could have been triggered

adamemerson commented:
@yehudasa I believe I have addressed all your review comments. If there's anything I missed/anything new you spot let me know.

I'm going to try shoving similar checks into the other operations.

yehudasa commented on src/rgw/rgw_op.h:
```diff
@@ -120,6 +120,37 @@ class RGWOp {
+  //
+  // The called function should set op_ret, similar to code in execute()
+  template<typename F>
+  void retry_raced_bucket_write(F&& f) {
```
@adamemerson RGWOp is not the place for a bucket specific operation. I suggest that you move it to a function in rgw_op.cc, make it return the ret value, and have f() return op_ret.

yehudasa commented on src/rgw/rgw_rados.cc:
```diff
@@ -11952,9 +11965,38 @@ int RGWRados::get_bucket_info(RGWObjectCtx& obj_ctx,
+               << "squirrelly, since this function should only be called "
+               << "in the event that an operation is rejected by "
+               << "version_check on the OSD." << dendl;
+    return -EALREADY;
```
do we want to even return an error here? is it an error? Maybe there was some kind of a forced  change that happened outside that ended up resetting the version. Whatever we read now should be fine, only need to warn.

adamemerson replied:
Can we have those kinds of forced changes? That is, is there any way where we can have the OSD tell that we have an invalid version, then we read direct from the OSD and it gives us the same version we had before? I would have figured that's an impossible situation, but if it can come up I'll remove the error return.

adamemerson replied:
Or at least I'd been working under the assumption that the versions must /always/ be monotonically increasing. I hadn't thought something external would be allowed to ratchet the version number down like that.

yehudasa replied:
The osd is not managing these versions, only verifies that they match (as requested by rgw). While not common, a version could be overriden (e.g., user injects object metadata).

adamemerson commented:
@yehudasa All right, I think I've got those things fixed up. I also added a quick cache expiry that @mattbenjamin , @ivancich  and I discussed yesterday.

adamemerson commented:
Also if anyone knows of any bucket-write operations I missed, please let me know. I hink I found all the ones in rgw_op.cc

yehudasa commented on src/common/options.cc:
```diff
@@ -5540,6 +5540,26 @@ std::vector<Option> get_rgw_options() {
+		     "cache are assumed stale and re-fetched. Zero is never.")
+    .add_tag("performance")
+    .add_service("rgw")
+    .set_long_description("The Rados Gateway stores metadata bout buckets in "
```
@adamemerson bout -> about

adamemerson commented:
@yehudasa Thtnka you for all the help and feedback.

tchaikov commented:
http://pulpito.ceph.com/kchai-2017-11-25_00:12:48-rgw-wip-kefu-testing-2017-11-25-0120-distro-basic-mira/

@adamemerson  is this ready to merge?

adamemerson commented:
@tchaikov Yes, sir! If it passes RGW tests it should be ready to merge.

mattbenjamin commented:
@adamemerson there are a number of failures;  a few are definite false positives, but there is a multisite test failure, at least one valgrind failure.

tchaikov commented:
@mattbenjamin the multisite test failure also exists in master. see http://pulpito.ceph.com/kchai-2017-11-26_02:57:34-rgw-master-distro-basic-mira/

mattbenjamin commented:
@adamemerson there are a number of failures;  a few are definite false positives, but there is a multisite test failure, at least one valgrind failure (that is a known issue, also reproduces on 12.2.2 and we are whitelisting it).

**http://qa-proxy.ceph.com/teuthology/kchai-2017-11-25_00:12:48-rgw-wip-kefu-testing-2017-11-25-0120-distro-basic-mira/1884703/teuthology.log

2017-11-25T02:34:45.720 INFO:tasks.rgw_multisite_tests:tasks.rgw_multi.tests.test_bucket_index_log_trim ... ok
2017-11-25T02:34:45.721 INFO:tasks.rgw_multisite_tests:
2017-11-25T02:34:45.721 INFO:tasks.rgw_multisite_tests:======================================================================
2017-11-25T02:34:45.721 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_multi_period_incremental_sync
2017-11-25T02:34:45.723 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T02:34:45.723 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2017-11-25T02:34:45.723 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2017-11-25T02:34:45.723 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2017-11-25T02:34:45.723 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 815, in test_multi_period_incremental_sync
2017-11-25T02:34:45.724 INFO:tasks.rgw_multisite_tests:    assert len(mdlog) == len(master_mdlog)
2017-11-25T02:34:45.724 INFO:tasks.rgw_multisite_tests:AssertionError
2017-11-25T02:34:45.724 INFO:tasks.rgw_multisite_tests:
2017-11-25T02:34:45.724 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T02:34:45.724 INFO:tasks.rgw_multisite_tests:Ran 21 tests in 6028.012s
2017-11-25T02:34:45.724 INFO:tasks.rgw_multisite_tests:
2017-11-25T02:34:45.725 INFO:tasks.rgw_multisite_tests:FAILED (failures=1)
2017-11-25T02:34:45.725 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):

http://pulpito.ceph.com/kchai-2017-11-25_00:12:48-rgw-wip-kefu-testing-2017-11-25-0120-distro-basic-mira/1884849/

2017-11-25T19:22:39.301 INFO:tasks.rgw_multisite_tests:tasks.rgw_multi.tests.test_bucket_index_log_trim ... FAIL
2017-11-25T19:22:39.301 INFO:tasks.rgw_multisite_tests:
2017-11-25T19:22:39.301 INFO:tasks.rgw_multisite_tests:======================================================================
2017-11-25T19:22:39.301 INFO:tasks.rgw_multisite_tests:ERROR: tasks.rgw_multi.tests.test_versioned_object_incremental_sync
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 697, in test_versioned_object_incremental_sync
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:    check_bucket_eq(source_conn, target_conn, bucket)
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 581, in check_bucket_eq
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:    return zone_conn2.check_bucket_eq(zone_conn1, bucket.name)
2017-11-25T19:22:39.302 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/zone_rados.py", line 75, in check_bucket_eq
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:    check_object_eq(k1, k2)
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/zone_rados.py", line 19, in check_object_eq
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:    eq(k1.get_contents_as_string(), k2.get_contents_as_string())
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/boto/s3/key.py", line 1782, in get_contents_as_string
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:    response_headers=response_headers)
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/boto/s3/key.py", line 1650, in get_contents_to_file
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:    response_headers=response_headers)
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/boto/s3/key.py", line 1482, in get_file
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:    query_args=None)
2017-11-25T19:22:39.303 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/boto/s3/key.py", line 1514, in _get_file_internal
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:    override_num_retries=override_num_retries)
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/boto/s3/key.py", line 343, in open
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:    override_num_retries=override_num_retries)
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/boto/s3/key.py", line 303, in open_read
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:    self.resp.reason, body)
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:S3ResponseError: S3ResponseError: 404 Not Found
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchKey</Code><BucketName>wbyhzq-16</BucketName><RequestId>tx000000000000000008587-005a19a9e4-1037-test-zone1</RequestId><HostId>1037-test-zone1-test-zonegroup</HostId></Error>
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:
2017-11-25T19:22:39.304 INFO:tasks.rgw_multisite_tests:======================================================================
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_multi_period_incremental_sync
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 815, in test_multi_period_incremental_sync
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:    assert len(mdlog) == len(master_mdlog)
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:AssertionError
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:
2017-11-25T19:22:39.305 INFO:tasks.rgw_multisite_tests:======================================================================
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_multipart_object_sync
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 1025, in test_multipart_object_sync
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:    zonegroup_bucket_checkpoint(zonegroup_conns, bucket.name)
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 398, in zonegroup_bucket_checkpoint
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:    zone_bucket_checkpoint(target_conn.zone, source_conn.zone, bucket_name)
2017-11-25T19:22:39.306 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 391, in zone_bucket_checkpoint
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:    (target_zone.name, source_zone.name, bucket_name)
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:AssertionError: finished bucket checkpoint for target_zone=test-zone3 source_zone=test-zone1 bucket=wbyhzq-54
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:======================================================================
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:FAIL: tasks.rgw_multi.tests.test_bucket_index_log_trim
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:Traceback (most recent call last):
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/virtualenv/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:    self.test(*self.arg)
2017-11-25T19:22:39.307 INFO:tasks.rgw_multisite_tests:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multi/tests.py", line 1111, in test_bucket_index_log_trim
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:    assert(len(cold_bilog) > 0)
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:AssertionError
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:----------------------------------------------------------------------
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:Ran 21 tests in 7877.440s
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:
2017-11-25T19:22:39.308 INFO:tasks.rgw_multisite_tests:FAILED (errors=1, failures=3)
2017-11-25T19:22:39.309 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
    manager.__enter__()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/task/__init__.py", line 121, in __enter__
    self.begin()
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multisite_tests.py", line 68, in begin
    raise RuntimeError('rgw multisite test failures')
RuntimeError: rgw multisite test failures
2017-11-25T19:22:39.342 ERROR:teuthology.run_tasks: Sentry event: http://sentry.ceph.com/sepia/teuthology/?q=03498395ad2544c8bad5e2d7d3a4dd59
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
    manager.__enter__()
  File "/home/teuthworker/src/git.ceph.com_git_teuthology_master/teuthology/task/__init__.py", line 121, in __enter__
    self.begin()
  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-kefu-testing-2017-11-25-0120/qa/tasks/rgw_multisite_tests.py", line 68, in begin
    raise RuntimeError('rgw multisite test failures')
RuntimeError: rgw multisite test failures
2017-11-25T19:22:39.342 DEBUG:teuthology.run_tasks:Unwinding manager rgw-multisite-tests
2017-11-25T19:22:39.503 DEBUG:teuthology.run_tasks:Unwinding manager rgw-multisite
2017-11-25T19:22:39.680 DEBUG:teuthology.run_tasks:Unwinding manager rgw**

mattbenjamin commented:
@cbodley ack?

adamemerson commented:
Fixes: #22517
